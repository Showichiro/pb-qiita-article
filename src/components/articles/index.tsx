import { Article } from "@/schemas";
import { FC } from "hono/jsx";

export const ArticlesTable: FC<{ articles: Array<Article> }> = ({
  articles,
}) => {
  return (
    <table class="table overflow-x-auto">
      <thead>
        <tr>
          <th>タイトル</th>
          <th>執筆者</th>
          <th>タグ</th>
          <th>いいね数</th>
          <th>ストック数</th>
        </tr>
      </thead>
      <tbody>
        {articles.map((article) => (
          <tr key={article.id} className="hover">
            <td>
              <a
                className="link"
                href={`https://qiita.com/${article.userId}/items/${article.id}`}
                target="_blank"
                rel="noopener noreferrer"
              >
                {article.title}
              </a>
            </td>
            <td>
              <a
                className="link"
                href={`https://qiita.com/${article.userId}`}
                target="_blank"
                rel="noopener noreferrer"
              >
                {article.userId}
                <span>{article.userName != "" && `(${article.userName})`}</span>
              </a>
            </td>
            <td>
              <ul>
                {article.tags.map((tag) => (
                  <li
                    key={`${article.id}-${tag.name}`}
                    className="badge badge-lg"
                  >
                    <a
                      href={`https://qiita.com/tags/${tag.name}`}
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      {tag.name}
                    </a>
                  </li>
                ))}
              </ul>
            </td>
            <td>{article.likesCount}</td>
            <td>{article.stocksCount}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};
