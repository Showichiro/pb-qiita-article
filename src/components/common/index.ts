export * from "./pagenation";
export * from "./header";
export * from "./pageTitle";
export * from "./layout";
export * from "./button";
export * from "./spinner";
