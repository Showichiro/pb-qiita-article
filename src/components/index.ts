export * from "./articles";
export * from "./common";
export * from "./ranking";
