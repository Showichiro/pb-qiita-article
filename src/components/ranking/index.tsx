import { ArticleCountGroupByUser, LikesCountSchema } from "@/schemas";
import { FC } from "hono/jsx";

export const Ranking: FC<{
  articleCountGroupByUsers: Array<ArticleCountGroupByUser>;
  likesCounts: Array<LikesCountSchema>;
}> = ({ articleCountGroupByUsers, likesCounts }) => {
  return (
    <>
      <h2>記事数ランキング</h2>
      <table>
        <thead>
          <tr>
            <th>順位</th>
            <th>執筆者</th>
            <th>記事数</th>
          </tr>
        </thead>
        <tbody>
          {articleCountGroupByUsers.map((user, index) => (
            <tr key={`article-count-${user.userId}`}>
              <td>{index + 1}</td>
              <td>
                <a href={`https://qiita.com/${user.userId}`} target="_blank">
                  {user.userId}
                  <span>{user.userName != "" && `(${user.userName})`}</span>
                </a>
              </td>
              <td>{user.count}</td>
            </tr>
          ))}
        </tbody>
      </table>
      <h2>いいね数ランキング</h2>
      <table>
        <thead>
          <tr>
            <th>順位</th>
            <th>執筆者</th>
            <th>いいね数</th>
          </tr>
        </thead>
        <tbody>
          {likesCounts.map((user, index) => (
            <tr key={`likes-count-${user.userId}`}>
              <td>{index + 1}</td>
              <td>
                <a href={`https://qiita.com/${user.userId}`}>
                  {user.userId}
                  <span>{user.userName != "" && `(${user.userName})`}</span>
                </a>
              </td>
              <td>{user.totalLikesCount}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
};

export const RankingRange: FC<{
  default: { since: string | null; until: string | null };
}> = ({ default: { since, until } }) => {
  return (
    <form action="/ranking" method="GET">
      <input type="date" name="sinse" value={since ?? undefined} />
      ～
      <input type="date" name="until" value={until ?? undefined} />
      <button type="submit">検索する</button>
    </form>
  );
};
