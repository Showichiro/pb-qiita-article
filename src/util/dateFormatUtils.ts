/**
 * @description Convert a Date object to a string in ISO 8601 format.
 * @param {Date} date The Date object to convert.
 * @returns {string} The ISO 8601 string representation of the Date object.
 * @example
 * const date = new Date();
 * const isoString = dateToDatetimeString(date);
 */
export const dateToDatetimeString = (date: Date): string => {
  return date.toISOString();
};

/**
 * @description Convert a datetime string to a date string.
 * @param {string} datetime The datetime string to convert.
 * @returns {string} The date string representation of the datetime string.
 * @example
 * const datetime = "2022-01-01T12:00:00.000Z";
 * const dateString = dateTimetoDateString(datetime);
 */
export const dateTimetoDateString = (datetime: string): string => {
  return new Date(datetime).toISOString().split("T")[0];
};
