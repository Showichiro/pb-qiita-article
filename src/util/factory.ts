import { schema } from "@/db";
import { renderer } from "@/renderer";
import { swaggerUI } from "@hono/swagger-ui";
import { OpenAPIHono } from "@hono/zod-openapi";
import { drizzle, DrizzleD1Database } from "drizzle-orm/d1";
import { cors } from "hono/cors";
import { logger } from "hono/logger";

export type Env = {
  Variables: {
    db: DrizzleD1Database<typeof schema>;
  };
  Bindings: {
    DB: D1Database;
  };
};

export const createHonoWithDBAndOpenAPI = () => {
  const app = new OpenAPIHono<Env>();
  // cors
  app.use(
    "*",
    cors({
      origin: ["http://localhost:5173"],
    })
  );
  // openapi
  app
    .doc("/doc", {
      openapi: "3.0.0",
      info: {
        version: "1.0.0",
        title: "PB Qiita API",
      },
      servers: [
        {
          url: "http://localhost:5173/",
          description: "local",
        },
        {
          url: "https://pb-qiita-articles.pages.dev/",
          description: "PRD",
        },
      ],
    })
    .use(logger())
    .use(renderer)
    .get(
      "/ui",
      swaggerUI({
        url: "/doc",
      })
    );
  app.use(async (c, next) => {
    const db = drizzle(c.env.DB, { schema, logger: true });
    c.set("db", db);
    await next();
  });
  return app;
};
