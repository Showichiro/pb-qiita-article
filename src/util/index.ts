export * from "./typeUtils";
export * from "./testUtils";
export * from "./dateFormatUtils";
export * from "./factory";
export * from "./dateParamsUtils";
