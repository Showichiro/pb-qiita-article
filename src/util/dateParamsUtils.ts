import { dateToDatetimeString } from "./dateFormatUtils";

export const processDateParam = (
  date: string | Date | null | undefined,
): string | null => {
  if (typeof date === "string") return date;
  if (date == null) return null;
  return dateToDatetimeString(date);
};
