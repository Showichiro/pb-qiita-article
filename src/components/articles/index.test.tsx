import { ArticlesTable, SelectBoxPageLimit } from "@/components";
import { renderer } from "@/util";

describe("ArticleTable", () => {
  it("should render articles html", async () => {
    const { text } = await renderer(
      <ArticlesTable
        articles={[
          {
            id: "123456",
            createdAt: new Date(0).toISOString(),
            likesCount: 1,
            stocksCount: 2,
            tags: [],
            title: "test-title",
            userId: "userId",
            userName: "userName",
          },
        ]}
      />
    );
    expect(text).toMatchSnapshot();
  });
});

describe("SelectBoxPageLimit", () => {
  it("should render selectbox page limit", async () => {
    const { text } = await renderer(<SelectBoxPageLimit limit={10} page={3} />);
    expect(text).toMatchSnapshot();
  });
});
