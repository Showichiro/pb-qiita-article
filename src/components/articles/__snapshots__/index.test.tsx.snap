// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticleTable > should render articles html 1`] = `"<table class="table overflow-x-auto"><thead><tr><th>タイトル</th><th>執筆者</th><th>タグ</th><th>いいね数</th><th>ストック数</th></tr></thead><tbody><tr class="hover"><td><a href="https://qiita.com/userId/items/123456" target="_blank" rel="noopener noreferrer" class="link">test-title</a></td><td><a href="https://qiita.com/userId" target="_blank" rel="noopener noreferrer" class="link">userId<span>(userName)</span></a></td><td><ul></ul></td><td>1</td><td>2</td></tr></tbody></table>"`;
